File: 3761-maximum-difference-between-even-and-odd-frequency-ii/maximum-difference-between-even-and-odd-frequency-ii.rb
# @param {String} s
# @param {Integer} k
# @return {Integer}
def max_difference(s, k)
  n = s.length
  pre_sum = Array.new(n + 1) { Array.new(5, 0) }

  n.times do |i|
    (0..4).each { |j| pre_sum[i + 1][j] = pre_sum[i][j] }
    pre_sum[i + 1][s[i].ord - '0'.ord] += 1
  end

  max = -Float::INFINITY

  (0..4).each do |a|
    (0..4).each do |b|
      next if a == b

      left = 0
      right = 0
      min = Array.new(2) { Array.new(2, Float::INFINITY) }

      while right < n
        while right - left + 1 >= k &&
              pre_sum[right + 1][a] > pre_sum[left][a] &&
              pre_sum[right + 1][b] > pre_sum[left][b]

          p = pre_sum[left][a] % 2
          q = pre_sum[left][b] % 2
          min[p][q] = [min[p][q], pre_sum[left][a] - pre_sum[left][b]].min
          left += 1
        end

        x = (pre_sum[right + 1][a] % 2) ^ 1
        y = pre_sum[right + 1][b] % 2
        max = [max, pre_sum[right + 1][a] - pre_sum[right + 1][b] - min[x][y]].max
        right += 1
      end
    end
  end

  max.to_i
end

File: 3761-maximum-difference-between-even-and-odd-frequency-ii/maximum-difference-between-even-and-odd-frequency-ii.go
func maxDifference(s string, k int) int {
	n := len(s);
	preSum := make([][]int, n+1);

	for i := range preSum {
		preSum[i] = make([]int, 5);
	}

	for i := 0; i < n; i++ {
		copy(preSum[i+1], preSum[i]);
		preSum[i+1][s[i]-'0']++;
	}

	maxDiff := -1 << 31 // tương đương Integer.MIN_VALUE

	for a := 0; a <= 4; a++ {
		for b := 0; b <= 4; b++ {
			if a == b {
				continue;
			}

			left, right := 0, 0;
			min := [2][2]int{};

			for i := 0; i < 2; i++ {
				for j := 0; j < 2; j++ {
					min[i][j] = 1 << 30; // tương đương Integer.MAX_VALUE / 2
				}
			}

			for right < n {
				for right-left+1 >= k &&
					preSum[right+1][a] > preSum[left][a] &&
					preSum[right+1][b] > preSum[left][b] {

					p := preSum[left][a] & 1;
					q := preSum[left][b] & 1;
					val := preSum[left][a] - preSum[left][b];
					if val < min[p][q] {
						min[p][q] = val;
					}

					left++;
				}

				x := (preSum[right+1][a] & 1) ^ 1;
				y := preSum[right+1][b] & 1;
				diff := preSum[right+1][a] - preSum[right+1][b] - min[x][y];
				if diff > maxDiff {
					maxDiff = diff;
				}
				right++;
			}
		}
	}

	return maxDiff;
}
